### ISSUE TITLE
C-state counters are missing in the Graphical Perfmon front end on Windows

### ISSUE THREAD

**Issue by henrikand** [opcm/pcm#347]
On my Windows machine `pcm.exe` shows:
```
C0 (active,non-halted) core residency: 5.07 %
C1 core residency: 13.72 %; 
C3 core residency: 0.30 %; 
C6 core residency: 0.00 %; 
C7 core residency: 80.91 %;

C0 package residency: 50.45 %;
C2 package residency: 25.83 %;
C3 package residency: 2.27 %;
C6 package residency: 0.02 %;
C7 package residency: 1.99 %;
C8 package residency: 19.44 %; 
C9 package residency: 0.00 %; 
C10 package residency: 0.00 %;
```
but when I use the Graphical Perfmon front end and view the new PCM* counters I can only see:

```
\PCM Core Counters(*)\core C0-state residency (%)
\PCM Core Counters(*)\core C3-state residency (%)
\PCM Core Counters(*)\core C6-state residency (%)
\PCM Core Counters(*)\core C7-state residency (%)

\PCM Socket Counters(*)\package C2-state residency (%)
\PCM Socket Counters(*)\package C3-state residency (%)
\PCM Socket Counters(*)\package C6-state residency (%)
\PCM Socket Counters(*)\package C7-state residency (%)
```

Hence, C1 core, C0 package, C8 package, C9 package and C10 package are missing.

I am using the third-party open source WinRing0 driver from RealTemp.

Based on code found in [PCMService.h](https://github.com/opcm/pcm/blob/2d2864e17f7b289b65d72334fc68bccaf08e02bd/PCM-Service_Win/PCMService.h#L493), it seems to be by design but I don't understand why some essential states have been removed.

**Comment by ogbrugge** [comment]
The PCM Windows service has bit-rotted a bit, recently there was another request for missing counters. I'll see what can be done to quickly fix this.

**Comment by henrikand** [comment]
Thanks for a quick reply. A fix in this area would be highly appreciated since I am trying to track down residence in these states and find it very convenient to base my study on Perfmon. The output from pcm.exe is a bit too wide and hard to parse for my purposes.

Let me know if you need help testing any early versions locally and I will gladly assist.

**Comment by ogbrugge** [comment]
I committed a patch to the internal repository, i managed to validate it on my laptop, now the c-states are available. Not sure when this is pushed to public github though, perhaps @opcm  can give an estimate.

If you need it quick, the patch basically adds the missing variables in the struct at the bottom of PCMService.h, further up creates them with CounterCreationData and adds them to the counterCollection, then adds them to the XXXCXStateResidencyHash_, then fills them with the getCoreCStateResidency and getPackageCstateResidency functions, calling these functions with the number of the missing cstate. The whole patch is copy and paste and then modify the numbers for the new cstates.

**Comment by henrikand** [comment]
Thanks! Looks like last commit is from Nov 3, 2021. Will see if I can follow your pattern and patch locally.

**Comment by opcm** [comment]
CPUs usually support only a subset of C-states. Using pcm-service and pcm are exclusive because they use the same CPU HW resources.

**Comment by ogbrugge** [comment]
If you stop the service, pcm works again. if not, just use as it says the -r option to clear the pmu

**Comment by henrikand** [comment]
It actually works for me (sometimes) to run both at the same time. Just FYI. Will try `-r` next time.

**Comment by henrikand** [comment]
Was able to implement the required changes listed by ogbrugge@ and it works.

However, the sum for C and PC states respectively does not add up to 100 (%) for me. The sum for each varies between 97 and 99. Assume that some sort of rounding is done that I am missing or I have done some mistake in my implementation.

The sum of all C and PC states is always exactly 100.0 using `pcm.exe`.

**Comment by ogbrugge** [comment]
Cool, you are welcome to submit a pull request for your change here so i can see if you missed something.

**Comment by henrikand** [comment]
Done @ogbrugge.

Just speculating but it could be that `PCMService` truncates all values and that the net effect is that the counters does not add up to exactly 100%.

**Comment by ogbrugge** [comment]
Not sure whether truncating is the cause of this, we're using 64 bit values for this. I need to check the code some more as to why this could happen.

**Comment by henrikand** [comment]
Great. 

I would still claim that what I think happens is that if `pcm.exe` shows

C0 package residency: 66.44 %; C2 package residency: 19.36 %; C3 package residency: 4.30 %; C6 package residency: 0.22 %; C7 package residency: 0.01 %; C8 package residency: 9.66 %; C9 package residency: 0.00 %; C10 package residency: 0.00 %;

PCMService converts this to 66, 19, 4, 0, 0, 9, 0 (sum equals 98).

**Comment by ogbrugge** [comment]
Right now I realize what you mean with truncating, it is basically rounding down due to converting from a float to an integer, I guess you change the counters to a floating point type if you want to experiment with it.

**Comment by henrikand** [comment]
Replacing `(__int64)` with `llround` for all C-state calculations fixes it for me locally.

E.g. `((PerformanceCounter^)CoreC0StateResidencyHash_[s])->RawValue = llround(100.*getCoreCStateResidency(0, oldSystemState, systemState));`

**Comment by ogbrugge** [comment]
Right the hard cast is bad, I’ve in addition tried changing the type to RawFraction instead of NumberOfInteger64 but I am running into issues now with not being able to start the service. It appears that for some reason the winpmem service is misbehaving.

**Comment by henrikand** [comment]
@ogbrugge: just a friendly status check. I can't see that you local changes have been landed yet. Is that correct? Also, it would be awesome with a correct rounding or even better with a transfer of a more high-precision format to ensure that all counters add up.

**Comment by ogbrugge** [comment]
@henrikand Correct, no change for this has landed yet, I am having severe difficulty getting the service to run on my laptop, I have no idea what I do wrong. I have submitted my patch as draft to the internal repository but I cannot test it. It used to work but no longer, and I cannot figure it out, probably something really easy but alas, windows is not very helpful with the error messages.

**Comment by henrikand** [comment]
@ogbrugge is it only the service that fails or can't you use `pcm.exe` either? I have seen issues as well, but for me, it has then helped to do: `PCM-Service.exe -Uninstall`, `PCM-Service.exe -Install`, `net start pcmservice`, restart device.  Also, I was never able to test-sign the `MSR.sys` driver and am using the third-party open source WinRing0 driver from RealTemp. Things are pretty stable for me on two different devices.

**Comment by ogbrugge** [comment]
Same here, using the winring0 driver, tried that about 20 times now, removed all pcmservice registry entries, rebooted several times but it always claims "the service is already running" and P'CM service stops with an error, event viewer says error %%526470 and more uninteresting stuff. pcm.exe works fine.

**Comment by henrikand** [comment]
Try checking the task manager to see if the service is actually running or not:

![PCM-Service](https://user-images.githubusercontent.com/8997299/144210339-4e4044dd-3e93-4365-85f9-f51836ef0170.png)

**Comment by ogbrugge** [comment]
I figured it out, it was the modifications i made to the service that was causing the problem... I fixed the code and it is running now but not properly tested.

**Comment by ogbrugge** [comment]
It just missed the push tonight, next time it will be in.

**Comment by opcm** [comment]
implemented in the master branch

**Comment by henrikand** [comment]
Tried the latest version and for me all package C states are zero. I have tried both Graphical Perfmon and a powershell script (which worked with the version from December). I do see valid values in core C-states.

**Comment by ogbrugge** [comment]
Odd, what CPU are you using? I did not see this in any of the comments.

I'll add some Package c-state counters to perfmon on my laptop to see if they show the right values or not.

**Comment by opcm** [comment]
is the pcm.exe command line utility itself showing non-zero package C-state residency?

**Comment by henrikand** [comment]
`Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz, 2592 Mhz, 6 Core(s), 12 Logical Processor(s)`

`pcm.exe` shows valid values (mainly ~100% PC0).

**Comment by henrikand** [comment]
Using same WinRing0 drivers as before. If I do `pcm-service -Uninstall` and reinstall the old service `PCM-Service.exe` things work again.

**Comment by ogbrugge** [comment]
Now that is odd, hmm perhaps there is a registry issue with how the counters are stored but the uninstall does not wipe those counters, only a restart of the service does that

**Comment by henrikand** [comment]
I both restarted the service and rebooted. Any hints on how I can "reset harder" and manually wipe the registry?

**Comment by ogbrugge** [comment]
Ok, so the change in the new pcm-service.exe is causing the package c-state counters to no longer work. Old pcm-service.exe was your version that you adapted with the pull request in this issue? Not the version that i had written in December?

**Comment by ogbrugge** [comment]
Manually wiping the registry means searching the registry for pcmservice or pcm-service and delete all entries you find with it but do not do it yet. I misunderstood you before, i missed the word old... so i thought you just reinstalled the service, that is why i wondered if it is registry that would be causing the problem but it is almost certain the different pcm-service executable.

**Comment by henrikand** [comment]
Yes. Sorry. Forgot to mention that. The version you had in December missed some states so I can't really compare with it but I don't recall that it gave zeros on all package C-states. I only did copy-paste changes.

**Comment by ogbrugge** [comment]
right, let me browse the change once more to see if i missed something obvious that would explain the missing c-state package counters

**Comment by ogbrugge** [comment]
I am able to reproduce the issue on my laptop, will need some time to investigate why this is not working as it should because I working on another feature atm.

**Comment by henrikand** [comment]
Thanks. Glad I had not messed something up. Would be great to be able to use the latest version instead of my hack.

**Comment by ogbrugge-work** [comment]
Issue was in hindsight easy to fix, will be in the next release but if you cant wait please look for the RawBase Counters for package c-states and change CoreCounter to SocketCounter, stupid copy and paste error that took some time to find. Should be about 16 or 18 replacements.

**Comment by henrikand** [comment]
I actually found that as well but can wait until the main repo is updated since my current build works OK. Would be great if you pinged this thread once the fix has landed.

Thanks!

**Comment by opcm** [comment]
this fix is now in the repo: https://github.com/opcm/pcm/commit/5f52f6291c43f591ba8a45b9ad6243b05a288e2d
